File: src/commonMain/resources/web/graal-file-viewer.js
"use strict";

// Graal's byte projections, shared by terrain, sheets, results and immutable refs.
const GraalFileViewer = (() => {
const LIMIT=1500000, states=new WeakMap();
const isCid=value=>typeof value==="string"&&/^sha256:[0-9a-f]{64}$/.test(value);
function fmtBytes(b){return b>1048576?(b/1048576).toFixed(1)+" MiB":b>1024?(b/1024).toFixed(1)+" KiB":b+" B";}
function esc(x){return (x??'').toString().replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;').replace(/"/g,'&quot;').replace(/'/g,'&#39;');}
const KW=/\b(fun|val|var|class|object|interface|override|suspend|return|if|else|when|for|while|import|package|private|public|internal|data|companion|def|lambda|function|const|let|new|static|void|int|long|boolean|public|final|try|catch|finally|throw|async|await)\b/g;
function codeView(text,maxLines){
  const lines=text.split('\n').slice(0,maxLines||1200);
  return '<div class="codeview">'+lines.map(line=>{
    let html='',at=0;
    const tokens=/"(?:[^"\\]|\\.)*"|'(?:[^'\\]|\\.)*'|\/\/.*$|#.*$|\b(?:fun|val|var|class|object|interface|override|suspend|return|if|else|when|for|while|import|package|private|public|internal|data|companion|def|lambda|function|const|let|new|static|void|int|long|boolean|final|try|catch|finally|throw|async|await)\b/g;
    for(const match of line.matchAll(tokens)){
      html+=esc(line.slice(at,match.index));
      const token=match[0],kind=/^(\/\/|#)/.test(token)?'com':/^["']/.test(token)?'str':'kw';
      html+='<span class="'+kind+'">'+esc(token)+'</span>';at=match.index+token.length;
    }
    return '<div>'+(html+esc(line.slice(at))||' ')+'</div>';
  }).join('')+'</div>';
}
function mdView(text){
  let h=esc(text);
  h=h.replace(/```([\s\S]*?)```/g,(m,c)=>'<pre>'+c+'</pre>');
  h=h.replace(/^###### (.*)$/gm,'<h3>$1</h3>').replace(/^##### (.*)$/gm,'<h3>$1</h3>').replace(/^#### (.*)$/gm,'<h3>$1</h3>')
     .replace(/^### (.*)$/gm,'<h3>$1</h3>').replace(/^## (.*)$/gm,'<h2>$1</h2>').replace(/^# (.*)$/gm,'<h1>$1</h1>');
  h=h.replace(/^\s*[-*] (.*)$/gm,'<li>$1</li>');
  h=h.replace(/\*\*([^*]+)\*\*/g,'<b>$1</b>').replace(/`([^`]+)`/g,'<code>$1</code>');
  h=h.replace(/\[([^\]]+)\]\(([^)]+)\)/g,(m,label,url)=>/^(https?:\/\/|\/(?!\/)|#)/i.test(url)?'<a href="'+url+'" target="_blank" rel="noopener">'+label+'</a>':label);
  h=h.replace(/\n\n+/g,'<br><br>');
  return '<div class="mdview">'+h+'</div>';
}
function classCard(b){
  if(b.byteLength<10)return null;
  const dv=b instanceof Uint8Array?new DataView(b.buffer,b.byteOffset,b.byteLength):new DataView(b);
  if(dv.getUint32(0)!==0xCAFEBABE)return null;
  const minor=dv.getUint16(4),major=dv.getUint16(6),pool=dv.getUint16(8);
  const jdk=major>=49?('Java '+(major-44)):'pre-1.5';
  return '<table class="classcard"><tr><td>magic</td><td>CAFEBABE</td></tr><tr><td>class file version</td><td>'+major+'.'+minor+' ('+jdk+')</td></tr><tr><td>constant pool</td><td>'+pool+' entries</td></tr><tr><td>size</td><td>'+fmtBytes(b.byteLength)+'</td></tr></table>';
}
function hexHead(b,n){
  const v=(b instanceof Uint8Array?b:new Uint8Array(b)).subarray(0,n||128);let out='';
  for(let i=0;i<v.length;i+=16){
    out+=Array.from(v.slice(i,i+16)).map(x=>x.toString(16).padStart(2,'0')).join(' ')+'   '+
         Array.from(v.slice(i,i+16)).map(x=>x>=32&&x<127?String.fromCharCode(x):'·').join('')+'\n';}
  return '<pre style="font-size:10px">'+esc(out)+'</pre><div style="color:var(--dim)">first '+v.length+' of '+fmtBytes(b.byteLength)+' — no friendlier projection for this type yet</div>';
}
function extOf(id){return id.split('.').pop().toLowerCase();}
/* ── shape strip: the Cascade Shape<S> of a text, RLE runs + fib ticks — sees content
      relationships with stochastic features. Same taxonomy as the Forge PWA strip. */
let _fib=null;function fibSet(){if(_fib)return _fib;_fib=new Set();let a=0,b=1;while(a<100000){_fib.add(a);const t=a+b;a=b;b=t;}return _fib;}
function classifyLine(l){
  const t=l.trim();
  if(!t)return '_';
  if(/^#{1,6} /.test(t)||/^=+$|^-{3,}$/.test(t))return 'H';
  if(/^\|.*\|/.test(t))return 'T';
  if(/^```/.test(t))return 'C';
  if(/^[-*+•] |^\d+\. /.test(t))return 'B';
  if(/^(fun|val|var|class|object|import|package|def|function|const|let|public|private|#include|@)/.test(t)||/[{};]\s*$/.test(t))return 'J';
  if(/^(##+ )?(section|chapter)\b/i.test(t))return 'S';
  return 'P';
}
function shapeRuns(text){
  const lines=text.split('\n');const runs=[];
  let c=null,n=0,start=1;
  lines.forEach((l,i)=>{const k=classifyLine(l);
    if(k===c)n++;else{if(c)runs.push({c,n,start});c=k;n=1;start=i+1;}});
  if(c)runs.push({c,n,start});
  return {runs,lines:lines.length};
}
const SHAPE_NAMES={H:'heading',S:'section',P:'prose','_':'blank',B:'bullet',T:'table',J:'code',C:'fence'};
function shapeStripHtml(text){
  const {runs,lines}=shapeRuns(text);const fib=fibSet();
  let depth=0;
  const cells=runs.map((r,i)=>{
    const isFib=fib.has(i);
    const title=(SHAPE_NAMES[r.c]||r.c)+' · '+r.n+' lines '+r.start+'–'+(r.start+r.n-1)+(isFib?' · depth '+i:'');
    return '<button class="shape-cell'+(isFib?' fib':'')+'" title="'+title+'" style="--c:var(--shape-'+r.c+',var(--shape-P));flex:'+r.n+' 0 2px" data-line="'+r.start+'"></button>';
  }).join('');
  const key=runs.map(r=>r.c).join('');
  return '<div class="shape-strip">'+cells+'</div><code class="shape-key">'+esc(key)+' · '+lines+' lines · '+runs.length+' runs</code>';
}

function element(tag,cls,text){const el=document.createElement(tag);if(cls)el.className=cls;if(text!=null)el.textContent=String(text);return el;}
function localUrl(value){const url=new URL(value,location.href);if(url.origin!==location.origin||!["http:","https:"].includes(url.protocol))throw Error("Non-local blob URL");return url.pathname+url.search;}
async function readBytes(response,limit=LIMIT){
  const reader=response.body?.getReader();if(!reader)throw Error("Response stream unavailable");
  let size=0;const chunks=[];
  try{
    if(Number(response.headers.get("content-length"))>limit)throw Error("Preview exceeds "+fmtBytes(limit));
    while(true){const {done,value}=await reader.read();if(done)break;size+=value.byteLength;if(size>limit)throw Error("Preview exceeds "+fmtBytes(limit));chunks.push(value);}
  }catch(error){await reader.cancel().catch(()=>{});throw error;}finally{reader.releaseLock();}
  const bytes=new Uint8Array(size);let offset=0;for(const chunk of chunks){bytes.set(chunk,offset);offset+=chunk.byteLength;}return bytes;
}
async function measure(bytes){
  if(!bytes.byteLength)return {raw:0,gzip:0,ratio:0};
  if(typeof CompressionStream==="undefined")return null;
  const compressed=await readBytes(new Response(new Blob([bytes]).stream().pipeThrough(new CompressionStream("gzip"))),LIMIT+65536);
  return {raw:bytes.byteLength,gzip:compressed.byteLength,ratio:compressed.byteLength/bytes.byteLength};
}
function kind(bytes,type="",id=""){
  const ext=extOf(id),ct=type.toLowerCase();
  if(ct.includes("java-vm")||ext==="class"||(bytes.length>=4&&bytes[0]===202&&bytes[1]===254&&bytes[2]===186&&bytes[3]===190))return "class";
  if(ct.startsWith("image/")||/^(svg|png|jpg|jpeg|gif|webp|avif)$/.test(ext))return "image";
  if(ext==="md"||ct.includes("markdown"))return "markdown";
  if(ct.includes("html")||/^(html|htm)$/.test(ext))return "html";
  if(ct.startsWith("text/")||/json|xml|javascript/.test(ct)||/^(kt|kts|java|py|js|ts|json|yaml|yml|css|sh|gradle|xml|txt|toml|properties|sql|rs|c|h|cpp|go|jsonl)$/.test(ext))return "text";
  try{const text=new TextDecoder("utf-8",{fatal:true}).decode(bytes);if(!/[\x00-\x08\x0e-\x1f]/.test(text))return "text";}catch(_){}
  return "binary";
}
function chip(cid,label="blob"){
  if(!isCid(cid))return esc(cid||"");
  return '<button class="blob-ref" data-blob-cid="'+cid+'" title="Open blob '+cid+'">'+esc(label)+' '+cid.slice(7,19)+'</button>';
}
function bindRefs(mount){mount.querySelectorAll("[data-blob-cid]").forEach(button=>button.addEventListener("click",e=>{e.stopPropagation();api.open({cid:button.dataset.blobCid});}));}
function dispose(mount){const state=states.get(mount);if(!state)return;state.controller.abort();for(const url of state.urls)URL.revokeObjectURL(url);states.delete(mount);}
async function sourceMates(id,mount,state){
  const region=element("div","file-mates");mount.append(region);
  try{
    const response=await fetch("/api/graal/decompile?source="+encodeURIComponent(id),{signal:state.controller.signal});
    const data=JSON.parse(new TextDecoder().decode(await readBytes(response)));if(state.controller.signal.aborted)return;
    if(!response.ok){region.textContent="Class projection unavailable: "+(data.error||response.status);return;}
    const mates=(data.mates||[]).slice(0,24);
    region.append(element("p","file-note",mates.length?"SourceFile mates: "+mates.length:"No SourceFile-mated class blob"));
    for(const mate of mates){
      const detail=element("details","mate"),title=element("summary",null,mate.className+" / "+(mate.exactRuntimeBlob?"exact runtime mate":mate.onClasspath?"classpath bytes differ":"not on runtime classpath"));
      detail.append(title);const body=element("div","matebody");
      body.innerHTML=chip(mate.blobCid)+(mate.runtimeCid?chip(mate.runtimeCid,"runtime"):"")+codeView(mate.decompiler?.pseudoSource||"No class projection",2400);
      detail.append(body);region.append(detail);
    }
    bindRefs(region);
  }catch(error){if(!state.controller.signal.aborted)region.textContent="Class projection unavailable: "+error.message;}
}
async function render(mount,options){
  dispose(mount);mount.replaceChildren();mount.classList.add("graal-file");
  const state={controller:new AbortController(),urls:[]};states.set(mount,state);
  const signal=state.controller.signal;
  if(options.signal?.aborted){state.controller.abort();return;}
  options.signal?.addEventListener("abort",()=>dispose(mount),{once:true});
  const status=element("p","file-note","Loading bytes");mount.append(status);
  try{
    let bytes=options.bytes,type=options.type||"",url=options.url&&localUrl(options.url);
    if(!bytes){
      let response;
      const urls=url?[url]:isCid(options.cid)?["/trikeshed/_cas/"+options.cid,"/api/lcnc/content?cid="+encodeURIComponent(options.cid)+"&view=bytes"]:[];
      if(!urls.length)throw Error("Blob identity required");
      for(const candidate of urls){response=await fetch(candidate,{signal});if(response.status!==404)break;}
      if(!response.ok)throw Error("Blob read failed: "+response.status);
      bytes=await readBytes(response);type=type||response.headers.get("content-type")||"";
    }
    bytes=bytes instanceof Uint8Array?bytes:new Uint8Array(bytes);
    if(bytes.byteLength>LIMIT)throw Error("Preview exceeds "+fmtBytes(LIMIT));
    if(signal.aborted)return;
    if(isCid(options.cid)){
      if(!crypto.subtle)throw Error("Content identity verification unavailable");
      const digest=await crypto.subtle.digest("SHA-256",bytes);
      const actual="sha256:"+Array.from(new Uint8Array(digest),b=>b.toString(16).padStart(2,"0")).join("");
      if(actual!==options.cid)throw Error("Content identity mismatch");
    }
    if(signal.aborted)return;
    const meta=element("div","file-meta");
    meta.innerHTML=(options.cid?chip(options.cid):"");meta.append(element("span",null,fmtBytes(bytes.byteLength)));
    const gauge=element("meter","kgauge");gauge.min=0;gauge.max=1;gauge.value=0;gauge.setAttribute("aria-label","Gzip compression ratio");
    const k=element("span","file-note","Measuring K");meta.append(gauge,k);mount.replaceChildren(meta);bindRefs(meta);
    measure(bytes).then(result=>{
      if(signal.aborted)return;
      if(!result){gauge.hidden=true;k.textContent="Gzip measurement unavailable";return;}
      gauge.value=Math.min(1,result.ratio);k.textContent=(result.ratio*100).toFixed(0)+"% K";
      meta.title="Gzip "+fmtBytes(result.gzip)+" / raw "+fmtBytes(result.raw)+". Compression proxy, not exact Kolmogorov complexity.";
    }).catch(()=>{if(!signal.aborted){gauge.hidden=true;k.textContent="Gzip measurement unavailable";}});
    const body=element("div","file-projection");mount.append(body);
    const mode=kind(bytes,type,options.id||""),text=()=>new TextDecoder().decode(bytes);
    if(mode==="image"){
      const ext=extOf(options.id||""),mime=type.startsWith("image/")?type:ext==="svg"?"image/svg+xml":"image/"+(ext==="jpg"?"jpeg":ext);
      const objectUrl=URL.createObjectURL(new Blob([bytes],{type:mime}));state.urls.push(objectUrl);
      const img=element("img","imgview");img.alt=options.id||options.cid||"Blob image";img.src=objectUrl;body.append(img);
    }else if(mode==="html"){
      const frame=element("iframe","htmlview");frame.title=options.id||"HTML blob";frame.setAttribute("sandbox","");
      frame.srcdoc='<meta http-equiv="Content-Security-Policy" content="default-src &#39;none&#39;; img-src data: blob:; style-src &#39;unsafe-inline&#39;">'+text();body.append(frame);
    }else if(mode==="class")body.innerHTML=classCard(bytes)||hexHead(bytes);
    else if(mode==="binary")body.innerHTML=hexHead(bytes);
    else{
      const value=text(),source=element("div");
      source.innerHTML=shapeStripHtml(value)+codeView(value);
      source.querySelectorAll("[data-line]").forEach(button=>button.addEventListener("click",()=>{
        const row=source.querySelector(".codeview")?.children[Number(button.dataset.line)-1];row?.scrollIntoView({block:"center"});
      }));
      if(mode==="markdown"){const rendered=element("div");rendered.innerHTML=mdView(value);body.append(rendered);const details=element("details");details.append(element("summary",null,"Source"),source);body.append(details);}
      else body.append(source);
      if(value.split("\n").length>1200)body.append(element("p","file-note","First 1200 lines; compression measures all loaded bytes."));
      if(options.sourceId&&/\.(kt|kts|java)$/i.test(options.id||""))await sourceMates(options.sourceId,body,state);
    }
  }catch(error){if(!signal.aborted)mount.replaceChildren(element("p","file-error",error.message));}
}
let dialog,dialogMount;
const api={isCid,kind,readBytes,measure,codeView,mdView,classCard,hexHead,shapeRuns,shapeStripHtml,render,dispose,chip,bindRefs,
  open(reference){
    if(!dialog){
      dialog=element("dialog","graal-file-dialog");const close=element("button","file-close","\u00d7");close.setAttribute("aria-label","Close file viewer");close.addEventListener("click",()=>dialog.close());
      dialog.append(close,element("h2"));dialogMount=element("div");dialog.append(dialogMount);document.body.append(dialog);
      dialog.addEventListener("close",()=>dispose(dialogMount));
    }
    dialog.querySelector("h2").textContent=reference.id||reference.cid||"File";
    if(!dialog.open)dialog.showModal();return render(dialogMount,reference);
  }
};
return api;
})();
if(typeof module!=="undefined")module.exports=GraalFileViewer;

File: src/jsHarness/kotlin/borg/trikeshed/web/harness/GraalFileViewer.kt
package borg.trikeshed.web.harness

import borg.trikeshed.graal.console.Projection
import kotlinx.browser.document
import kotlinx.browser.window
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.await
import kotlinx.coroutines.promise
import org.khronos.webgl.ArrayBuffer
import org.khronos.webgl.Uint8Array
import org.khronos.webgl.get
import org.w3c.dom.Element
import org.w3c.dom.HTMLElement
import org.w3c.dom.get
import org.w3c.dom.url.URL
import org.w3c.fetch.Response
import org.w3c.files.Blob
import org.w3c.files.BlobPropertyBag
import kotlin.js.Promise

/** graal-file-viewer.js: byte projections of a blob into a mount element. Published as `window.GraalFileViewer`. */
@OptIn(DelicateCoroutinesApi::class)
object GraalFileViewer {
    private class State(val controller: AbortController, val urls: ArrayList<String> = ArrayList())

    private val states: dynamic = js("new WeakMap()")

    private fun fmtBytes(b: Double) = Projection.fmtBytesViewer(b)

    private fun ByteArrayOf(bytes: Uint8Array): ByteArray = Int8Array(bytes.buffer, bytes.byteOffset, bytes.length).unsafeCast<ByteArray>()

    private fun localUrl(value: String): String {
        val url = URL(value, window.location.href)
        if (url.origin != window.location.origin || (url.protocol != "http:" && url.protocol != "https:")) throw Error("Non-local blob URL")
        return url.pathname + url.search
    }

    suspend fun readBytes(response: Response, limit: Int = FileViewerRules.LIMIT): Uint8Array {
        val reader: dynamic = response.body?.asDynamic()?.getReader() ?: throw Error("Response stream unavailable")
        var size = 0
        val chunks = ArrayList<Uint8Array>()
        try {
            if (jsNumber(response.headers.get("content-length")) > limit) throw Error("Preview exceeds " + fmtBytes(limit.toDouble()))
            while (true) {
                val part: dynamic = (reader.read() as Promise<dynamic>).await()
                if (part.done == true) break
                val value = part.value.unsafeCast<Uint8Array>()
                size += value.byteLength
                if (size > limit) throw Error("Preview exceeds " + fmtBytes(limit.toDouble()))
                chunks.add(value)
            }
        } catch (error: Throwable) {
            try { (reader.cancel() as Promise<dynamic>).await() } catch (_: Throwable) {}
            throw error
        } finally {
            reader.releaseLock()
        }
        val bytes = Uint8Array(size)
        var offset = 0
        for (chunk in chunks) { bytes.set(chunk, offset); offset += chunk.byteLength }
        return bytes
    }

    private class Measure(val raw: Int, val gzip: Int, val ratio: Double)

    private suspend fun measure(bytes: Uint8Array): Measure? {
        if (bytes.byteLength == 0) return Measure(0, 0, 0.0)
        if (jsTypeOf(window.asDynamic().CompressionStream) == "undefined") return null
        val stream: dynamic = Blob(arrayOf(bytes)).asDynamic().stream().pipeThrough(js("new CompressionStream('gzip')"))
        val compressed = readBytes(Response(stream), FileViewerRules.LIMIT + 65536)
        return Measure(bytes.byteLength, compressed.byteLength, compressed.byteLength.toDouble() / bytes.byteLength)
    }

    private fun decodeFatal(bytes: Uint8Array): String? = try {
        js("new TextDecoder('utf-8',{fatal:true})").decode(bytes) as String
    } catch (_: Throwable) { null }

    private fun decode(bytes: Uint8Array): String = js("new TextDecoder()").decode(bytes) as String

    private fun head(bytes: Uint8Array): IntArray = IntArray(minOf(4, bytes.length)) { bytes[it].toInt() and 0xFF }

    fun kind(bytes: Uint8Array, type: String = "", id: String = ""): String =
        FileViewerRules.kind(head(bytes), type, id) { decodeFatal(bytes) }

    fun chip(cid: String?, label: String = "blob"): String = FileViewerRules.chip(cid, label)

    fun bindRefs(mount: Element, reference: dynamic = null) {
        val buttons = mount.querySelectorAll("[data-blob-cid]")
        for (i in 0 until buttons.length) {
            val button = buttons.item(i).unsafeCast<HTMLElement>()
            button.addEventListener("click", { e ->
                e.stopPropagation()
                // Framed archive content has a document CID but is restored through its manifest.
                val cid = button.dataset["blobCid"]
                if (reference != null && reference.cid == cid) {
                    val copy: dynamic = js("Object.assign({}, reference)"); copy.signal = undefined; open(copy)
                } else { val r: dynamic = jsObject(); r.cid = cid; open(r) }
            })
        }
    }

    fun dispose(mount: Element) {
        val state = states.get(mount).unsafeCast<State?>() ?: return
        state.controller.abort()
        for (url in state.urls) URL.revokeObjectURL(url)
        states.delete(mount)
    }

    private suspend fun fetchLimited(url: String, state: State, limit: Int): Pair<Response, dynamic> {
        val response = fetchResponse(url, requestInit(signal = state.controller.signal))
        val data: dynamic = JSON.parse(decode(readBytes(response, limit)))
        return response to data
    }

    private suspend fun sourceMates(id: String, mount: Element, state: State) {
        val region = el("div", "file-mates"); mount.append(region)
        try {
            val (response, data) = fetchLimited("/api/graal/decompile?source=" + encodeURIComponent(id), state, FileViewerRules.LIMIT)
            if (state.controller.signal.aborted) return
            if (!response.ok) { region.textContent = "Class projection unavailable: " + (if (truthy(data.error)) data.error else response.status); return }
            val mates = ((data.mates ?: js("[]")) as Array<dynamic>).take(24)
            region.append(el("p", "file-note", if (mates.isNotEmpty()) "SourceFile mates: " + mates.size else "No SourceFile-mated class blob"))
            for (mate in mates) {
                val detail = el("details", "mate")
                val title = el("summary", null, FileViewerRules.mateTitle(jsString(mate.className), truthy(mate.exactRuntimeBlob), truthy(mate.onClasspath)))
                detail.append(title)
                val body = el("div", "matebody")
                body.innerHTML = chip(mate.blobCid as String?) + (if (truthy(mate.runtimeCid)) chip(mate.runtimeCid as String?, "runtime") else "") +
                    Projection.codeViewTokens(text(mate.decompiler?.pseudoSource).ifEmpty { "No class projection" }, 2400)
                detail.append(body); region.append(detail)
            }
            bindRefs(region)
        } catch (error: Throwable) {
            if (!state.controller.signal.aborted) region.textContent = "Class projection unavailable: " + errorMessage(error)
        }
    }

    private fun text(v: dynamic): String = if (truthy(v)) jsString(v) else ""

    private suspend fun classProjection(id: String, mount: Element, state: State) {
        val region = el("section", "classdetail"); region.append(el("p", "file-note", "Loading class projection")); mount.append(region)
        try {
            val (response, data) = fetchLimited("/api/graal/classfile?id=" + encodeURIComponent(id), state, 2097152)
            if (state.controller.signal.aborted) return
            if (!response.ok || truthy(data.error)) {
                region.replaceKids(el("p", "file-error", "Class projection unavailable: " + (if (truthy(data.error)) data.error else response.status))); return
            }
            val methods = if (js("Array").isArray(data.methods) as Boolean) (data.methods as Array<dynamic>).take(80) else emptyList()
            val fields = if (js("Array").isArray(data.fields) as Boolean) (data.fields as Array<dynamic>).take(40) else emptyList()
            val interfaces = if (js("Array").isArray(data.interfaces) as Boolean) (data.interfaces as Array<dynamic>).joinToString(", ") { jsString(it) } else ""
            region.innerHTML = FileViewerRules.classHeading(text(data.className).ifEmpty { text(data.id) }.ifEmpty { id }) + FileViewerRules.projectionTable(listOf(
                "super" to text(data.superClass),
                "interfaces" to interfaces,
                "source" to text(data.sourceFile),
                "classpath" to FileViewerRules.tristate(data.onClasspath, "present", "absent"),
                "exact runtime blob" to FileViewerRules.tristate(data.exactRuntimeBlob, "true", "false"),
                "api" to text(data.classfileApi).ifEmpty { "java.lang.classfile" },
            ))
            if (fields.isNotEmpty()) {
                val table = el("table", "class-members")
                table.innerHTML = FileViewerRules.fieldRows(fields.map { text(it.name) to text(it.descriptor) })
                region.append(table)
            }
            if (methods.isNotEmpty()) {
                val table = el("table", "class-members")
                table.innerHTML = FileViewerRules.methodRows(methods.map {
                    Triple(text(it.name), text(it.descriptor), if (js("Array").isArray(it.instructions) as Boolean) jsString(it.instructions.length) else text(it.instructionCount))
                })
                region.append(table)
            }
        } catch (error: Throwable) {
            if (!state.controller.signal.aborted) region.replaceKids(el("p", "file-error", "Class projection unavailable: " + errorMessage(error)))
        }
    }

    suspend fun render(mount: HTMLElement, options: dynamic) {
        dispose(mount); mount.replaceKids(); mount.classList.add("graal-file")
        val state = State(AbortController())
        states.set(mount, state)
        val signal = state.controller.signal
        val outer = options.signal.unsafeCast<AbortSignal?>()
        if (outer?.aborted == true) { state.controller.abort(); return }
        outer?.addEventListener("abort", { dispose(mount) }, js("({once:true})"))
        val status = el("p", "file-note", "Loading bytes"); mount.append(status)
        try {
            var bytes: dynamic = options.bytes
            var type: String = text(options.type)
            val url: String? = if (text(options.url).isNotEmpty()) localUrl(options.url as String) else null
            if (bytes == null) {
                var response: Response? = null
                val cid = options.cid as String?
                val urls = if (url != null) listOf(url) else if (FileViewerRules.isCid(cid)) listOf("/trikeshed/_cas/$cid", "/api/lcnc/content?cid=" + encodeURIComponent(cid!!) + "&view=bytes") else emptyList()
                if (urls.isEmpty()) throw Error("Blob identity required")
                for (candidate in urls) { response = fetchResponse(candidate, requestInit(signal = signal)); if (response.status.toInt() != 404) break }
                if (!response!!.ok) throw Error("Blob read failed: " + response.status)
                bytes = readBytes(response)
                if (type.isEmpty()) type = response.headers.get("content-type") ?: ""
            }
            val data: Uint8Array = if (bytes is Uint8Array) bytes else Uint8Array(bytes.unsafeCast<ArrayBuffer>())
            if (data.byteLength > FileViewerRules.LIMIT) throw Error("Preview exceeds " + fmtBytes(FileViewerRules.LIMIT.toDouble()))
            if (signal.aborted) return
            if (FileViewerRules.isCid(options.cid as String?)) {
                if (window.asDynamic().crypto?.subtle == null) throw Error("Content identity verification unavailable")
                val actual = sha256(data)
                if (actual != options.cid) throw Error("Content identity mismatch")
            }
            if (signal.aborted) return
            val meta = el("div", "file-meta")
            meta.innerHTML = if (text(options.cid).isNotEmpty()) chip(options.cid as String?) else ""
            meta.append(el("span", null, fmtBytes(data.byteLength.toDouble())))
            val gauge = el("meter", "kgauge"); val g: dynamic = gauge; g.min = 0; g.max = 1; g.value = 0
            gauge.setAttribute("aria-label", "Gzip compression ratio")
            val k = el("span", "file-note", "Measuring K")
            meta.append(gauge, k); mount.replaceKids(meta); bindRefs(meta, options)
            GlobalScope.promise {
                try {
                    val result = measure(data)
                    if (signal.aborted) return@promise
                    if (result == null) { gauge.hidden = true; k.textContent = "Gzip measurement unavailable"; return@promise }
                    g.value = minOf(1.0, result.ratio)
                    k.textContent = (js("(function(v){return (v*100).toFixed(0)})")(result.ratio) as String) + "% K"
                    meta.title = "Gzip " + fmtBytes(result.gzip.toDouble()) + " / raw " + fmtBytes(result.raw.toDouble()) + ". Compression proxy, not exact Kolmogorov complexity."
                } catch (_: Throwable) {
                    if (!signal.aborted) { gauge.hidden = true; k.textContent = "Gzip measurement unavailable" }
                }
            }
            val body = el("div", "file-projection"); mount.append(body)
            val id = text(options.id)
            val mode = kind(data, type, id)
            when (mode) {
                "image" -> {
                    val objectUrl = URL.createObjectURL(Blob(arrayOf(data), BlobPropertyBag(type = FileViewerRules.imageMime(type, id))))
                    state.urls.add(objectUrl)
                    val img = el("img", "imgview"); val i: dynamic = img
                    i.alt = id.ifEmpty { text(options.cid) }.ifEmpty { "Blob image" }; i.src = objectUrl; body.append(img)
                }
                "html" -> {
                    val frame = el("iframe", "htmlview"); frame.title = id.ifEmpty { "HTML blob" }; frame.setAttribute("sandbox", "")
                    frame.asDynamic().srcdoc = FileViewerRules.HTML_CSP + decode(data); body.append(frame)
                }
                "class" -> {
                    val raw = ByteArrayOf(data)
                    body.innerHTML = Projection.classCard(raw, ::fmtBytes) ?: Projection.hexHead(raw, 128, ::fmtBytes, Projection::escAll)
                    if (id.isNotEmpty()) classProjection(id, body, state)
                }
                "binary" -> body.innerHTML = Projection.hexHead(ByteArrayOf(data), 128, ::fmtBytes, Projection::escAll)
                else -> {
                    val value = decode(data)
                    val source = el("div")
                    source.innerHTML = Projection.shapeStripHtml(value, false, Projection::escAll) + Projection.codeViewTokens(value)
                    val cells = source.querySelectorAll("[data-line]")
                    for (c in 0 until cells.length) {
                        val button = cells.item(c).unsafeCast<HTMLElement>()
                        button.addEventListener("click", {
                            val row = source.querySelector(".codeview")?.children?.item(jsNumber(button.dataset["line"]).toInt() - 1)
                            row?.asDynamic()?.scrollIntoView(js("({block:'center'})"))
                        })
                    }
                    if (mode == "markdown") {
                        val rendered = el("div"); rendered.innerHTML = Projection.mdViewSafe(value); body.append(rendered)
                        val details = el("details"); details.append(el("summary", null, "Source"), source); body.append(details)
                    } else body.append(source)
                    if (value.split("\n").size > FileViewerRules.TEXT_LINES) body.append(el("p", "file-note", "First 1200 lines; compression measures all loaded bytes."))
                    if (text(options.sourceId).isNotEmpty() && FileViewerRules.jvmSource(id)) sourceMates(options.sourceId as String, body, state)
                }
            }
        } catch (error: Throwable) {
            if (!signal.aborted) mount.replaceKids(el("p", "file-error", errorMessage(error)))
        }
    }

    private var dialog: HTMLElement? = null
    private var dialogMount: HTMLElement? = null

    fun open(reference: dynamic): Promise<Unit> {
        if (dialog == null) {
            val d = el("dialog", "graal-file-dialog")
            val close = el("button", "file-close", "\u00d7"); close.setAttribute("aria-label", "Close file viewer")
            close.addEventListener("click", { d.asDynamic().close() })
            d.append(close, el("h2")); val m = el("div"); d.append(m); document.body!!.append(d)
            d.addEventListener("close", { dispose(m) })
            dialog = d; dialogMount = m
        }
        val d = dialog!!
        d.querySelector("h2")!!.textContent = text(reference.id).ifEmpty { text(reference.cid) }.ifEmpty { "File" }
        if (d.asDynamic().open != true) d.asDynamic().showModal()
        return GlobalScope.promise { render(dialogMount!!, reference) }
    }

    /** window.GraalFileViewer — the page-wide handle other bundles and the landscape call. */
    fun install() {
        if (window.asDynamic().GraalFileViewer != null) return
        val api: dynamic = jsObject()
        api.isCid = { v: dynamic -> jsTypeOf(v) == "string" && FileViewerRules.isCid(v as String) }
        api.kind = { b: Uint8Array, t: String?, i: String? -> kind(b, t ?: "", i ?: "") }
        api.readBytes = { r: Response, limit: Int? -> GlobalScope.promise { readBytes(r, limit ?: FileViewerRules.LIMIT) } }
        api.codeView = { t: String, n: Int? -> Projection.codeViewTokens(t, n ?: 1200) }
        api.mdView = { t: String -> Projection.mdViewSafe(t) }
        api.shapeStripHtml = { t: String -> Projection.shapeStripHtml(t, false, Projection::escAll) }
        api.render = { m: HTMLElement, o: dynamic -> GlobalScope.promise { render(m, o) } }
        api.dispose = { m: Element -> dispose(m) }
        api.chip = { c: String?, l: String? -> chip(c, l ?: "blob") }
        api.bindRefs = { m: Element, r: dynamic -> bindRefs(m, r) }
        api.open = { r: dynamic -> open(r) }
        window.asDynamic().GraalFileViewer = api
    }
}

external fun encodeURIComponent(s: String): String

@JsName("Int8Array")
private external class Int8Array(buffer: ArrayBuffer, byteOffset: Int, length: Int)

suspend fun sha256(bytes: Uint8Array): String {
    val digest = (window.asDynamic().crypto.subtle.digest("SHA-256", bytes) as Promise<ArrayBuffer>).await()
    val view = Uint8Array(digest)
    val sb = StringBuilder("sha256:")
    for (i in 0 until view.length) sb.append((view[i].toInt() and 0xFF).toString(16).padStart(2, '0'))
    return sb.toString()
}
